crates/wasi/src/preview2/tcp.rs

```diff
@@ -208,21 +267,36 @@ impl HostOutputStream for TcpWriteStream {
                     return Ok(());
                 }
 
-                Err(e) => return Err(StreamError::LastOperationFailed(e.into())),
+                Err(e) if e.kind() == std::io::ErrorKind::BrokenPipe => {
+                    return Err(StreamError::Closed);
+                }
+
+                Err(e) => {
+                    return Err(StreamError::LastOperationFailed(e.into()));
+                }
             }
         }
 
         Ok(())
     }
 
     fn flush(&mut self) -> Result<(), StreamError> {
+        if self.inner.is_shut_down_for_sending() {
+            return Err(StreamError::Closed);
+        }
+
         // `flush` is a no-op here, as we're not managing any internal buffer. Additionally,
         // `write_ready` will join the background write task if it's active, so following `flush`
         // with `write_ready` will have the desired effect.
         Ok(())
     }
 
     fn check_write(&mut self) -> Result<usize, StreamError> {
+        if self.inner.is_shut_down_for_sending() {
+            self.last_write = LastWrite::Done;
+            return Err(StreamError::Closed);
+        }
```

Comment:
Should this be checked after checking for `Waiting` below? It's a bit weird to have a pending write and then call shutdown, but I think the `shutdown` syscall will cause the background write to wake up and report its failure?

Reply:
The user-facing behavior implemented in this PR seems correct to me. I.e. to return `StreamError::Closed` even if an background write was still in progress.

With the current solution it's the responsibility of the caller to properly wait for all data to be fully acknowledged/flushed _before_ calling `shutdown`.

This issue is not limited to just `shutdown`, but also `drop`.

I think the true question is: whether or not to abandon the background write (as currently current implemented) or detach the running background tasks and let the background write finish up even after the WASI input stream has been closed (on `drop` or `shutdown`).

Reply:
Oh I'm not advocating for letting this proceed in the background, I'm more thinking that it might be best to shuffle this to get the response from the background task before checking the shutdown status. I'm under the impression that a `shutdown` will cancel pending writes by marking the socket as writable and returning `Ok(0)` or `EPIPE` or something like that. That would mean that after the `shutdown` it's not an indefinite block for the background task to finish up and for us to get the result.

Reply:
> it might be best to shuffle this to get the response from the background task before checking the shutdown status.

What would be the benefit? On line 296 I reset the last_write status which drops the `AbortOnDropJoinHandle`. So the background task will always be aborted, right?

Reply:
It's true yeah that largely the same thing will happen, but I'd at least personally prefer to explicitly clean up the task rather than let it get cleaned up in the background if possible. It's sort of six-to-one-half-dozen-or-the-other though.

I'm not certain of the portability of the behavior of marking the socket as writable and cancelling all writes once `shutdown` is issued though. If that's not the actual OS behavior then switching the order here wouldn't work.

Given the discussion [here](https://bytecodealliance.zulipchat.com/#narrow/stream/219900-wasi/topic/What.20happens.20to.20in-progress.20background.20writes.20on.20drop.3F/near/411819385) though I'm also a bit wary of cancelling active writes since that could show up as surprising for users, so maybe I should be advocating for letting this proceed in the background...